Implementing slugs for routing for members and characters.

Pyrodius0 opened the pull request:
Implemented slugs for user and for characters.

This solves the bug where character routing would break.

Added database\hacks\character/memberSlugConversion.php files to track what needs to be executed to convert slugs.

fixes #5 
resolves #5 
closes #5

Lemmings19 commented on database/migrations/2020_08_18_032818_add_member_character_slug.php:
```diff
@@ -0,0 +1,36 @@
+        });
+
+        Schema::table('members', function (Blueprint $table) {
+            $table->string('slug')->nullable()->after('username');
```
We're going to use this slug to lookup members from the database, so I think we should index both of these new slugs.

The slug function in PHP also limits the slug's length to 50 characters, and that's loads for a username. I think we can also set these fields to be 50 characters for faster indexing+lookup (I am sort of assuming it's faster).

These fields are also required to load a page for either member or character, so we shouldn't allow null on them. (`username` and `name` also have NOT NULL set to true)

```
$table->string('slug', 50)->after('username')->index();
```

Lemmings19 replied:
Guilds and instances have their slugs set to 255 characters, so I think there's an opportunity to optimize those fields too at some point.

Pyrodius0 replied:
updated as per your feedback.

Pyrodius0 commented:
> This looks really good, thanks for doing this! I'm waiting for the migration to get sorted out before I run it locally and give it a test. (see comment attached to the migration)
> 
> Other than that, I don't see any issues!

I updated the migration as requested. I updated the same migration  and wiped my database and remigrated to test everything. 

Looks good.

Lemmings19 commented on database/migrations/2020_08_18_032818_add_member_character_slug.php:
```diff
@@ -0,0 +1,40 @@
+     */
+    public function down()
+    {
+        Schema::table('characters', function (Blueprint $table) {
+            //
+        });
+
+        Schema::table('members', function (Blueprint $table) {
+            //
+        });
```
Just a note that these don't do anything unless we program them to undo the changes we made. (but we can skip it for now)

`php artisan migrate:rollback` can then be run to undo a migration.

Pyrodius0 replied:
yeah, I was just stubbing it for later if we need.

Lemmings19 commented:
Trying to figure out how to edit my review or open a new one. Ignore any weird stuff regarding reviews.

Lemmings19 commented:
Need to update character and member links in the dashboard: https://github.com/thatsmybis/thatsmybis/blob/master/resources/views/dashboard.blade.php#L26

Lemmings19 commented on resources/views/member/show.blade.php:
```diff
@@ -55,7 +55,7 @@
                                     <small class="text-muted">
                                         on
-                                        <a href="{{route('character.show', ['guildSlug' => $guild->slug, 'name' => $character->name]) }}"
+                                        <a href="{{route('character.show', ['guildSlug' => $guild->slug, 'nameSlug' => $character->slug]) }}"
```
Looks like an existing bug, but `$character->slug` needs to be changed to `$itemCharacter->slug`

Pyrodius0 replied:
Updated

Lemmings19 commented:
Two changes needed (dashboard, `itemCharacter`), then it should be good for a merge.

Pyrodius0 commented:
> Need to update character and member links in the dashboard: https://github.com/thatsmybis/thatsmybis/blob/master/resources/views/dashboard.blade.php#L26

Not sure what we are missing, but I already updated that. 

![image](https://user-images.githubusercontent.com/69526309/90581761-ddb65900-e180-11ea-8e89-c4cae26fa990.png)

Pyrodius0 commented on resources/views/dashboard.blade.php:
```diff
@@ -23,14 +23,14 @@
                         <li class="bg-lightest mt-3 mb-3 p-3">
                             <h2>
-                                <a href="{{ route('member.show', ['guildSlug' => $member->guild->slug, 'username' => $member->username]) }}" class="text-uncommon font-weight-medium">
+                                <a href="{{ route('member.show', ['guildSlug' => $member->guild->slug, 'usernameSlug' => $member->slug]) }}" class="text-uncommon font-weight-medium">
```
@Lemmings19  I think you were looking at the wrong branch.

Pyrodius0 commented on resources/views/dashboard.blade.php:
```diff
@@ -23,14 +23,14 @@
                                 @foreach ($member->characters as $character)
                                     <li class="list-inline-item bg-tag rounded pt-0 pl-2 pb-1 pr-2 m-2">
-                                        <a href="{{route('character.show', ['guildSlug' => $member->guild->slug, 'name' => $character->name]) }}"
+                                        <a href="{{route('character.show', ['guildSlug' => $member->guild->slug, 'nameSlug' => $character->slug]) }}"
```
@Lemmings19 I think you were looking at the wrong branch.

Lemmings19 commented:
You're right, I probably was on the wrong branch or was missing some changes.

Looks good!

:shipit:
